src/Compilers/Core/Portable/Emit/EditAndContinue/DeletedTypeDefinition.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Runtime.InteropServices;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis.Emit.EditAndContinue
+{
+    /// <summary>
+    /// Represents a type referenced from a deleted member (as distinct from a type that has been deleted)
+    /// </summary>
+    internal class DeletedTypeDefinition : ITypeDefinition
+    {
+        [return: NotNullIfNotNull("typeReference")]
+        public static ITypeReference? TryCreate(ITypeReference? typeReference, Dictionary<ITypeDefinition, DeletedTypeDefinition> cache)
+        {
+            if (typeReference is ITypeDefinition typeDef)
+            {
+                if (!cache.TryGetValue(typeDef, out var deletedType))
+                {
+                    deletedType = new DeletedTypeDefinition(typeDef);
+                    cache.Add(typeDef, deletedType);
+                }
+
+                return deletedType;
+            }
+
+            return typeReference;
+        }
+
+        private readonly ITypeDefinition _oldTypeReference;
```

Comment:
>_oldTypeReference [](http://example.com/codeflow?start=41&length=17)

It looks like we're delegating to `_oldTypeReference` in most cases. I'm curious where the behavior is different using a `DeletedTypeDefinition` rather than simply using `_oldTypeReference` directly.

Reply:
The `AsDefinition` method is the most important, as without that various lookups fail, and things end up trying to add to table indexes after they've been finalized, but also Aleksey said it was important to wrap everything so I'm wrapping everything :)